postings.go

```diff
@@ -45,21 +45,44 @@ func (p *memPostings) get(name, value string) Postings {
 	return newListPostings(l)
 }
 
-var allLabel = labels.Label{}
+var allPostingsKey = labels.Label{}
 
 // add adds a document to the index. The caller has to ensure that no
 // term argument appears twice.
 func (p *memPostings) add(id uint64, lset labels.Labels) {
 	p.mtx.Lock()
 
 	for _, l := range lset {
-		p.m[l] = append(p.m[l], id)
+		p.addFor(id, l)
 	}
-	p.m[allLabel] = append(p.m[allLabel], id)
+	p.addFor(id, allPostingsKey)
 
 	p.mtx.Unlock()
 }
 
+func (p *memPostings) addFor(id uint64, l labels.Label) {
+	list := append(p.m[l], id)
+	p.m[l] = list
+
+	// There is no guarantee that no higher ID was inserted before as they may
+	// be generated independently before adding them to postings.
+	// We repair order violations on insert. The invariant is that the first n-1
+	// items in the list are already sorted.
+	for i := len(list) - 1; i >= 1; i-- {
```

Comment:
There is still a race here, but I might be wrong. Also this case is a long shot, but hear me out :)

Let the existing list be: [1, 2, 3]
Now we insert: 4, 5, 6, 7 concurrently in this way:
[1, 2, 3, 6, 4, 5, 7].

Now, if the check for `5` runs before check for `4`, which means in the end, we end up with: `[1, 2, 3, 4, 6, 5, 7]`

Reply:
Also, if we have `[1, 2, 3]` and insert 6 and then 4 and then 5.
We get at one point: 
```
list := [1, 2, 3, 6, 4, 5]
p.m[l] = list
```

Even the swapping for `6, 4` that happens is not visible to the new list as it is a copy, AFAICS.

Reply:
The insertion itself is guarded by mutex, so there is no concurrency at play here.
Before each insert all elements so far are sorted. Just the last one is potentially out of order.

Reply:
Aah, got it. LGTM!